### PR TITLE
[tune] Ensure arguments passed to tune `remote_run` match

### LINKED ISSUE #18655: [Bug] Ray Client ray.tune.run fails on remote workers with max_concurrent_trials TypeError
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Tune

### What happened + What you expected to happen

My `ray.tune.run` succeeds on a local workstation. 

In a gcp cluster `ray.tune.run` fails with the following error:

```
2021-09-15 21:01:56.201 INFO ray.worker Using address ray://10.128.0.36:10001 set in the environment variable RAY_ADDRESS
Traceback (most recent call last):
  File "/home/max/anaconda3/envs/torch/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/max/anaconda3/envs/torch/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/max/src/research/es/tune4.py", line 339, in <module>
    main()
  File "/home/max/src/research/es/tune4.py", line 335, in main
    fn(symbol, options)
  File "/home/max/src/research/es/tune4.py", line 189, in tune
    analysis = ray.tune.run(
  File "/home/max/anaconda3/envs/torch/lib/python3.8/site-packages/ray/tune/tune.py", line 317, in run
    return ray.get(
  File "/home/max/anaconda3/envs/torch/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 88, in wrapper
    return getattr(ray, func.__name__)(*args, **kwargs)
  File "/home/max/anaconda3/envs/torch/lib/python3.8/site-packages/ray/util/client/api.py", line 42, in get
    return self.worker.get(vals, timeout=timeout)
  File "/home/max/anaconda3/envs/torch/lib/python3.8/site-packages/ray/util/client/worker.py", line 238, in get
    res = self._get(to_get, op_timeout)
  File "/home/max/anaconda3/envs/torch/lib/python3.8/site-packages/ray/util/client/worker.py", line 265, in _get
    raise err
types.RayTaskError(TypeError): ray::run() (pid=3578, ip=10.128.0.36)
  File "/home/max/anaconda3/envs/torch/lib/python3.8/site-packages/ray/tune/tune.py", line 475, in run
    if max_concurrent_trials < 1:
TypeError: '<' not supported between instances of 'list' and 'int'
```

My code doesn't specify `max_concurrent_trials` when calling `ray.tune.run`:
```
analysis = ray.tune.run(
    ray.tune.with_parameters(train, inputs=(...)),
    metric=..., mode=...,
    config=config,
    scheduler=MedianStoppingRule(),
    search_alg=HyperOptSearch(),
    num_samples=10,
)
```

Where does that `max_concurrent_trials` of type `list` come from?

Software versions:
* Ray: https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp38-cp38-manylinux2014_x86_64.whl fetched on 2021-09-15 20:30 UTC.
* Python: 3.8.10
* Ubuntu 20.04.3 LTS

### Reproduction script

TBD

### Anything else

This error doesn't happen when `ray.tune.run` is invoked with no head node, or on a head node with `RAY_ADDRESS=auto`.

This error only happens with Ray Client use case https://docs.ray.io/en/master/cluster/ray-client.html when I set `RAY_ADDRESS=ray://10.128.0.36:10001` and run my script from elsewhere.

It suggests that IPC between a remote cluster Ray Client and the head node may be broken.

@richardliaw 

### Are you willing to submit a PR?

I would if I had a clue.

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by Yard1**
## Why are these changes needed?

Ensures that the arguments passed to `remote_run` inside `tune.run` match exactly with those in `run`. Adds a test for a remote call with a `search_alg` defined.

We can use something like `inspect` here to ensure that only the actual argument names are selected from the `locals()` dict, but I felt that would add unnecessary complexity, while the problem can be solved by simply copying the `locals()` dict before any other code in the function is ran, as it is done in the PR. 

## Related issue number

Closes https://github.com/ray-project/ray/issues/18655

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Yard1** [comment]
BTW. IMO we should consider this way of handling this in other cases where we are doing a remote function. It eliminates the possibility of mismatched arguments, leading to bugs as in the linked issue

File: python/ray/tune/tune.py [hunk 1]
```diff
@@ -299,6 +299,10 @@ def run(
         TuneError: Any trials failed and `raise_on_failed_trial` is True.
     """
 
+    # at this point, locals is equal to args and kwargs
```
Comment (xwjiang2010) [inline python/ray/tune/tune.py:302]: Can you beef it up a bit so that people don't accidentally add stuff prior to this line.

**Review by xwjiang2010 — APPROVED** [review]

File: python/ray/tune/tune.py [hunk 2]
```diff
@@ -314,52 +321,9 @@ def run(
...
-                ray_auto_init,
-                run_errored_only,
-                global_checkpoint_period,
-                with_server,
-                upload_dir,
-                sync_to_cloud,
-                sync_to_driver,
-                sync_on_checkpoint,
-                _remote=False))
+        return ray.get(remote_run.remote(_remote=False, **remote_run_kwargs))
```
[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/tune/tune.py:324]: I wanted to do that, but I was unsure how (with `run` being a method). Any ideas how?

[the remark this answers is not in the document]
Comment (Yard1) [reply python/ray/tune/tune.py:317]: Added a test!

**Review by krfricke — APPROVED** [review]
Thanks!

**Pull request merged**